frostsnap_core/src/device.rs

```diff
@@ -279,58 +412,144 @@ impl FrostSigner {
 
                 let agg_nonces = (0..n_signatures_requested).map(|i| sign_req.agg_nonce(i));
 
-                self.action_state = Some(SignerState::AwaitingSignAck {
-                    key_id,
+                self.action_state = Some(SignerState::AwaitingSignAck(Box::new(AwaitingSignAck {
+                    root_key,
+                    access_structure_id,
+                    encrypted_secret_share: encrypted_secret_share.ciphertext,
+                    my_party_index: party_index,
                     sign_task: checked_sign_task.clone(),
                     agg_nonces: agg_nonces.collect(),
                     parties: sign_req.parties().collect(),
                     nonce_start_index: my_nonces.start,
                     nonces_remaining: my_nonces.nonces_remaining,
-                });
-                Ok(vec![DeviceSend::ToUser(
+                    coord_share_decryption_contrib: sign_req.coord_share_decryption_contrib,
+                })));
+                Ok(vec![DeviceSend::ToUser(Box::new(
                     DeviceToUserMessage::SignatureRequest {
                         sign_task: checked_sign_task,
-                        key_id,
+                        appkey,
                     },
-                )])
+                ))])
             }
-            (None, CoordinatorToDeviceMessage::DisplayBackup { key_id }) => {
-                let _key = self
-                    .secret_shares
-                    .get(&key_id)
-                    .ok_or(Error::signer_invalid_message(
-                        &message,
-                        "signer doesn't have a share for this key",
-                    ))?;
+            (
+                None,
+                CoordinatorToDeviceMessage::DisplayBackup {
+                    key_id,
+                    access_structure_id,
+                    coord_share_decryption_contrib,
+                    party_index,
```

Comment:
One idea for later is that users could press down to cycle through each backup belonging to an access structure.
Another opportunity to add a magical animation. Could be nicer for the app, one less selection menu of boring share indicies.

Though if you're only intending to backup just one share at a time, you could have some very minor security concern of someone recording your screen. We do also want to encourage geographical separation of backups as a principle so maybe it is not a good idea, but in this case they are already one one device.

Reply:
We talked about this on call. I'm sure a good design will come out of our discussions.